### PR TITLE
client: support split/scatter region request

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

ref https://github.com/tikv/pd/issues/2965

### What is changed and how it works?

support pdclient for split/scatter regions. see https://github.com/pingcap/kvproto/pull/686

### Check List

Tests

- Unit test

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -96,7 +96,9 @@ type Client interface {
 	ScatterRegion(ctx context.Context, regionID uint64) error
 	// ScatterRegionWithOption scatters the specified region with the given options, should use it
 	// for a batch of regions.
-	ScatterRegionWithOption(ctx context.Context, regionID uint64, opts ...ScatterRegionOption) error
+	ScatterRegionWithOption(ctx context.Context, regionID uint64, opts ...RegionOption) error
```
Comment (rleungx) [inline client/client.go:99]: Is there possible to add this option to ScatterRegion instead of creating a new method?

Comment (Yisaer) [reply client/client.go:99]: The option for Scatter and Split is much similar. If there is no repo using this api, I tend to rename it as `RegionOption`.

@overvenus Is `ScatterRegionWithOption` being used in tools?

Comment (Yisaer) [reply client/client.go]: updated.

File: client/client.go [hunk 2]
```diff
@@ -928,30 +925,53 @@ func (c *client) GetOperator(ctx context.Context, regionID uint64) (*pdpb.GetOpe
...
 
 	ctx, cancel := context.WithTimeout(ctx, c.timeout)
 	resp, err := c.leaderClient().ScatterRegion(ctx, &pdpb.ScatterRegionRequest{
-		Header:   c.requestHeader(),
-		RegionId: regionID,
-		Group:    group,
+		Header:     c.requestHeader(),
+		Group:      group,
+		RegionsId:  regionsID,
+		RetryLimit: uint64(retryLimit),
```
Comment (rleungx) [inline client/client.go:965]: Is there possible not to do the conversion?

File: pkg/testutil/testutil.go [hunk 3]
```diff
@@ -44,6 +44,18 @@ func WaitUntil(c *check.C, f CheckFunc) {
 	c.Fatal("wait timeout")
 }
 
+// WaitUntilByInterval repeatedly evaluates f() for a given period of time, util it returns true.
```
Comment (rleungx) [inline pkg/testutil/testutil.go:47]: It is almost the same as WaitUntil

Comment (Yisaer) [reply pkg/testutil/testutil.go:40]: updated.

File: pkg/testutil/testutil.go [hunk 4]
```diff
@@ -34,12 +34,17 @@ type CheckFunc func(c *check.C) bool
 
 // WaitUntil repeatedly evaluates f() for a period of time, util it returns true.
 func WaitUntil(c *check.C, f CheckFunc) {
```
Comment (rleungx) [inline pkg/testutil/testutil.go:36]: Maybe we can add an option for it and we don't need two functions.

Comment (Yisaer) [reply pkg/testutil/testutil.go:36]: I think the option is not necessary here.

Comment (rleungx) [reply pkg/testutil/testutil.go:36]: We don't need to add WaitUntilByInterval since these two functions do almost the same things

Comment (Yisaer) [reply pkg/testutil/testutil.go]: updated.

File: client/client.go [hunk 5]
```diff
@@ -892,24 +897,16 @@ func (c *client) UpdateServiceGCSafePoint(ctx context.Context, serviceID string,
...
+		span = opentracing.StartSpan("pdclient.ScatterRegions", opentracing.ChildOf(span.Context()))
 		defer span.Finish()
 	}
-	options := &ScatterRegionOp{}
+	options := &RegionsOp{}
 	for _, opt := range opts {
 		opt(options)
 	}
-	return c.scatterRegionsWithGroup(ctx, regionID, options.group)
+	return c.scatterRegionsWithOptions(ctx, regionsID, options.group, options.retryLimit)
```
Comment (rleungx) [inline client/client.go:909]:
```suggestion
	return c.scatterRegionsWithOptions(ctx, regionsID, options)
```

File: client/client.go [hunk 6]
```diff
@@ -928,30 +925,53 @@ func (c *client) GetOperator(ctx context.Context, regionID uint64) (*pdpb.GetOpe
...
 }
 
+// SplitRegions split regions by given split keys
+func (c *client) SplitRegions(ctx context.Context, splitKeys [][]byte, opts ...RegionsOption) (*pdpb.SplitRegionsResponse, error) {
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("pdclient.SplitRegions", opentracing.ChildOf(span.Context()))
+		defer span.Finish()
+	}
+	start := time.Now()
+	defer func() { cmdDurationGetOperator.Observe(time.Since(start).Seconds()) }()
```
Comment (rleungx) [inline client/client.go:935]: cmdDurationGetOperator？

**Review by rleungx — APPROVED** [review]

File: go.mod [hunk 7]
```diff
@@ -52,6 +52,7 @@ require (
 )
 
 replace (
+	github.com/pingcap/kvproto => github.com/Yisaer/kvproto v0.0.0-20201110052545-128289baa6a4
```
Comment (HunDunDM) [inline go.mod:55]: There are some [comments](https://github.com/pingcap/kvproto/pull/686) on `kvproto`.

File: server/grpc_service.go [hunk 8]
```diff
@@ -712,10 +732,6 @@ func (s *Server) ScatterRegion(ctx context.Context, request *pdpb.ScatterRegionR
 		region = core.NewRegionInfo(request.GetRegion(), request.GetLeader())
 	}
 
-	if rc.IsRegionHot(region) {
```
Comment (HunDunDM) [inline server/grpc_service.go:715]: Please explain the reason for canceling this judgment.

File: server/schedule/region_scatterer.go [hunk 9]
```diff
@@ -241,7 +241,7 @@ func (r *RegionScatterer) ScatterRegions(regions map[uint64]*core.RegionInfo, fa
 		retryLimit = maxRetryLimit
 	}
 	ops := make([]*operator.Operator, 0, len(regions))
-	for currentRetry := 0; currentRetry < retryLimit; currentRetry++ {
+	for currentRetry := 0; currentRetry < retryLimit+1; currentRetry++ {
```
Comment (HunDunDM) [inline server/schedule/region_scatterer.go:244]:
```suggestion
	for currentRetry := 0; currentRetry <= retryLimit; currentRetry++ {
```

**Review by HunDunDM — CHANGES_REQUESTED** [review]
Need to update kvproto

File: server/grpc_service.go (hunk 8 above)
Comment (Yisaer) [reply server/grpc_service.go:715]: https://github.com/tikv/pd/blob/3c625632b06c956ab74ca6e6a2941f6888200071/server/schedule/region_scatterer.go#L270-L287

duplicated with these checks.

**Review by disksing — APPROVED** [review]

File: server/grpc_service.go [hunk 10]
```diff
@@ -704,18 +704,36 @@ func (s *Server) ScatterRegion(ctx context.Context, request *pdpb.ScatterRegionR
...
 	}
 
+	if len(request.GetRegionsId()) > 0 {
+		ops, failures, err := rc.GetRegionScatter().ScatterRegionsByID(request.GetRegionsId(), request.GetGroup(), int(request.GetRetryLimit()))
+		if err != nil {
+			return nil, err
+		}
+		for _, op := range ops {
+			if ok := rc.GetOperatorController().AddOperator(op); !ok {
+				failures[op.RegionID()] = fmt.Errorf("region %v failed to add operator", op.RegionID())
```
Comment (rleungx) [inline server/grpc_service.go:714]: How to handle these error messages?

Comment (Yisaer) [reply server/grpc_service.go:714]: adding a log debug to help debug.

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**